File: py/data/vehicle_state.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from data import vehicle_state_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class VehicleState:
    x: float = 0.0
    y: float = 0.0
    heading: float = 0.0
    u: float = 0.0
    v: float = 0.0
    
    _proto_message: vehicle_state_pb2.VehicleState = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = vehicle_state_pb2.VehicleState()

    def _populate_proto(self):
        self._proto_message.x = self.x
        self._proto_message.y = self.y
        self._proto_message.heading = self.heading
        self._proto_message.u = self.u
        self._proto_message.v = self.v

    @classmethod
    def _from_proto(cls, proto_message) -> "VehicleState":
        return cls(
            x = proto_message.x,
            y = proto_message.y,
            heading = proto_message.heading,
            u = proto_message.u,
            v = proto_message.v,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "VehicleState":
        message = vehicle_state_pb2.VehicleState()
        message.ParseFromString(data)
        return cls(
            x = message.x,
            y = message.y,
            heading = message.heading,
            u = message.u,
            v = message.v,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "VehicleState":
        message = vehicle_state_pb2.VehicleState()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/data/vehicle_state.cpp
#include "data/vehicle_state.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Data
{
VehicleState::VehicleState(const PbData::VehicleState& protobuf) {
    x = protobuf.x();
    y = protobuf.y();
    heading = protobuf.heading();
    u = protobuf.u();
    v = protobuf.v();
}

VehicleState::operator PbData::VehicleState() const {
    PbData::VehicleState ret;
    ret.set_x(x);
    ret.set_y(y);
    ret.set_heading(heading);
    ret.set_u(u);
    ret.set_v(v);
    return ret;
}

std::string VehicleState::serializeAsJsonString() const {
    PbData::VehicleState protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string VehicleState::serializeAsProtobufString() const {
    PbData::VehicleState protobuf(*this);
    return protobuf.SerializeAsString();
}

bool VehicleState::deserializeFromJsonString(const std::string& str) {
    PbData::VehicleState protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool VehicleState::deserializeFromProtobufString(const std::string& str) {
    PbData::VehicleState protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}